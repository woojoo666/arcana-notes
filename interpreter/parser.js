
// --------------------------- PARSER ----------------------------

// comparison of parser libraries: https://tomassetti.me/parsing-in-javascript/
// For now I am using the nearley.js library to generate the parser, as Canopy (http://canopy.jcoglan.com) doesn't support left-recursive rules
//     and Ohm(https://github.com/harc/ohm) makes it complicated to work with rules that have multiple "arity" (see https://nextjournal.com/dubroy/ohm-parsing-made-easy)
// I chose Nearley because it looks simple and concise. The playground (omrelli.ug/nearley-playground/) made it very easy to test grammars as well.

// first: convert indentation into INDENT and DEDENT, convert newlines to commas, and ignore indentation and newlines in braced blocks.
//        convert words and operators into tokens
//
// Then parse everything using Nearley.js
// Everything in the AST should correspond to an "actor" node, which we will use for the reactive updating.
// Even if-statements, for-loops, operators, all will use "actor" nodes

// Later I might switch from Nearley to a custom parser that will allow me to parse blocks independently to localize errors.


// note: `nearley` and `grammar` objects included by the scripts in the <head> of the html

function parse (text) {  // TODO: when we write our tokenizer, change this to take in tokens instead of raw text
	console.log('parsing: \n' + text);
	try {
		const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
		parser.feed(text);
		return parser.results[0]; // return first successful parsing
	} catch (err) {
		console.log("Error at character " + err.offset);
	}
}
