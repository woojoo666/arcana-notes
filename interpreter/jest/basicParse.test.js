
import { parse } from '../parser.js';

// todo: add test for all basic syntax, eg unary operators, templates, parameters, @-blocks, cloning, if-else, etc
test('booleans', () => {
	const testCode = 'x: true';
	let parsed = parse(testCode);
	console.log(JSON.stringify(parsed));
    expect(parsed.statements[0].value.type).toEqual('boolean');
    expect(parsed.statements[0].value.value).toEqual('true');
});

test('strings', () => {
	expect(parse('x: ""').statements[0].value.type).toEqual('string');
	expect(parse('x: "test"').statements[0].value.value).toEqual('\"test\"'); // notice that the AST preserves the original code of all primitives, including strings
	expect(parse('x: "pekora be like \\"ha↑ha↑ha↑\\""').statements[0].value.type).toEqual('string'); // test escaped quotes and special characters
});

test('"undefined" literal', () => {
	expect(parse('x: undefined').statements[0].value.type).toEqual('undefined');
});

test('numeric keys', () => {
    const testCode = '5: 888';
	let parsed = parse(testCode);
    expect(parsed.statements[0].key).toEqual('5');
});

// note: boolean keys are useful for implementing ternaries and conditionals
test('boolean keys', () => {
	const testCode = 'true: 888';
	let parsed = parse(testCode);
    expect(parsed.statements[0].key).toEqual('true');
});

test('property access', () => {
    expect(parse('bar: x.foo').statements[0].value.type).toEqual('memberAccess');
    expect(parse('bar: (foo: 100).foo').statements[0].value.type).toEqual('memberAccess'); // inline property access
    expect(parse('bar: foo(x: 100).x').statements[0].value.type).toEqual('memberAccess');  // inline clone + property access
});

test('computed property access', () => {
	let code = 'a: b[c]';
	let parsed = parse(code);

	expect(parsed).toMatchSnapshot();
	expect(parsed.statements[0].value.type).toEqual('memberAccess');
	expect(parsed.statements[0].value.key.type).toEqual('reference');

	let code2 = 'a: b[c + max(d,e)]';
	let parsed2 = parse(code2);
	expect(parsed2).toMatchSnapshot();
	expect(parsed2.statements[0].value.key.operator).toEqual('+');
	expect(parsed2.statements[0].value.key.right.type).toEqual('clone');

    expect(parse('bar: (x: 100)["x"]').statements[0].value.type).toEqual('memberAccess');    // inline computed property access
    expect(parse('bar: foo(x: 100)["x"]').statements[0].value.type).toEqual('memberAccess'); // inline clone + computed property access
});
