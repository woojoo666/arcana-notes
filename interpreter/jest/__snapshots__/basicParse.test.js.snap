// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computed property access 1`] = `
Object {
  "parameters": null,
  "statements": Array [
    Object {
      "key": "a",
      "type": "property",
      "value": Object {
        "key": Object {
          "name": "c",
          "type": "reference",
        },
        "source": Object {
          "name": "b",
          "type": "reference",
        },
        "type": "memberAccess",
      },
    },
  ],
  "type": "block",
}
`;

exports[`computed property access 2`] = `
Object {
  "parameters": null,
  "statements": Array [
    Object {
      "key": "a",
      "type": "property",
      "value": Object {
        "key": Object {
          "left": Object {
            "name": "c",
            "type": "reference",
          },
          "operator": "+",
          "right": Object {
            "block": Object {
              "parameters": null,
              "statements": Array [
                Object {
                  "items": Array [
                    Array [
                      Object {
                        "name": "d",
                        "type": "reference",
                      },
                    ],
                  ],
                  "type": "subList",
                },
                Object {
                  "items": Array [
                    Array [
                      Object {
                        "name": "e",
                        "type": "reference",
                      },
                    ],
                  ],
                  "type": "subList",
                },
              ],
              "type": "block",
            },
            "source": Object {
              "name": "max",
              "type": "reference",
            },
            "type": "clone",
          },
          "type": "binop",
        },
        "source": Object {
          "name": "b",
          "type": "reference",
        },
        "type": "memberAccess",
      },
    },
  ],
  "type": "block",
}
`;
